src/Agent.Plugins/TfsVCCliManager.cs

```diff
@@ -172,6 +182,24 @@ protected async Task<string> RunPorcelainCommandAsync(FormatFlags formatFlags, p
             return string.Join(Environment.NewLine, result.Output ?? new List<string>());
         }
 
+        protected async Task<TfsVCPorcelainCommandResult> TryRunPorcelainCommandAsync(FormatFlags formatFlags, int retriesOnFailure, params string[] args)
+        {
+            var result = await TryRunPorcelainCommandAsync(formatFlags, args);
+            int attempt = 0;
+            while (attempt < retriesOnFailure && result.Exception != null)
+            {
+                result = await TryRunPorcelainCommandAsync(formatFlags, args);
+                if (result.Exception != null)
+                {
+                    int sleep = Math.Min(200 * (int)Math.Pow(5, attempt), 30000);
+                    ExecutionContext.Output($"Sleeping for {sleep} ms");
+                    await Task.Delay(sleep);
+                };
+                attempt++;
+            };
+            return result;
+        }
+
```

Comment:
It wasn't totally clear to me that this will work for this case since we're returning a result. What does the returned object represent here? It looks like it should contain any output data. Are we going to need data from previous runs included? This won't do that.

Reply:
@damccorm  No, we don't need it.
The result object contains two fields: Error and Output. If retry was successful we need only the information which contains in Output field in last, successful attempt. Otherwise we should return result which contains only the latest error. ( in all retries the probably should be the same)

Reply:
My understanding (and this might not be right) was that in these retry situations, we might partially succeed and make it part of the way through before failing - in that case, there might be valid information in the first failed attempt's output as well as the second successful attempt's output.

How are we using this output? Is it just getting logged out to the user or is it being used elsewhere by us somehow? If its just getting logged out I'm not too worried about it.

Reply:
@RuslanRasulovAkv I think I'd like this question answered before we merge this one - other than that I feel good about it

Reply:
@damccorm  Yes, of course, today I will check it more closely

Reply:
@damccorm  In error case we just log out the output so we don't use it anywhere else.

Reply:
@RuslanRasulovAkv I'm not really worried about the failure case, I'm worried about the case where the first attempt partially succeeds but fails before completion and the second case succeeds.

For example - lets say you're trying to get 3 workspaces A, B, and C and you have the following sequence of attempts:

1) Successfully gets A and B, fails on C
2) Successfully gets C.

Will the output from retry 2 contain all the information that we need? I see we eventually serialize this output into some workspace objects - https://github.com/microsoft/azure-pipelines-agent/blob/8db185bb973d9684c96057e664ad4561ee7419e9/src/Agent.Plugins/TeeCliManager.cs#L302. I haven't dug in further than that, but presumably we use those somewhere.

Reply:
Yes, we may have appropriate workspace in any retry (But we can't check where, and also i am not sure that tfvc return workspaces one by one). After we get  the workspaces (in TfsVCSourceProvider it is an Array of objects) agents try to use existing one if it possible ( that is why we need all workspaces their for checking).

Reply:
Ok. I think I'm ok with merging this now because its probably strictly better than the existing code. If we see issues with partial successes after this, we should revisit this